File: qp-fbug/qp-fbug-0.1/src/chrome/content/util/statemachine.js
var EXPORTED_SYMBOLS = ["loadModule"];
loadModule = function(QPFBUG)
{

with (QPFBUG.Classes){
with (Lang){

    var owner = QPFBUG.Classes;

    //--------------------------------- StateMachine --------------------------------
    // This object works like a state machine and is usually used for keeping
    // the state of an object or component.
    // A transition has a name, from and to values.
    owner.StateMachine =
        function(){

            // []String states
            // String initial
            // [][3]String transitions
            var constructor = function(states, initial, transitions){
                // non of the values can be null
                assert(states);
                assert(initial);
                assert(transitions);
                // the initial state should be one of the states.
                assert(states.indexOf(initial)!=-1);
                //[]String states: state names
                this.states = states;
                //String   initial: initial state name
                this.initial = initial;

                //map<String,[3]String> transitions: all transitions
                this.transitions = {};
                for (i = 0; i < transitions.length; i++) {
                    //both from and to states should be in the list of states
                    assert(states.indexOf(transitions[i][1]!=-1));
                    assert(states.indexOf(transitions[i][2])!=-1);
                    this.transitions[transitions[i][0]] = transitions[i];
                }

                //String: current state
                this.state = initial;
                //boolean: it is true if state machine is in transition
                this.inTransition = false;
                //String
                this.currentTransition = null;
                //long: last change time
                this.lastChangeTime = currentTimeMillis();
            };

            constructor.prototype = {

                //String transition: transition name
                startTransit: function(transition)
                {
                    assert(this.transitions[transition]);
                    assert(this.state == this.transitions[transition][1], this.state);
                    this.inTransition = true;
                    this.state = null;
                    this.currentTransition = transition;
                    this.lastChangeTime = currentTimeMillis();
                },

                endTransit: function()
                {
                    assert(this.inTransition);
                    this.state = this.transitions[this.currentTransition][2];
                    this.inTransition = false;
                    this.currentTransition = null;
                    this.lastChangeTime = currentTimeMillis();
                },

                doTransit : function(transition)
                {
                    this.startTransit(transition);
                    this.endTransit();
                },

                getState : function() {
                    return this.state;
                },

                isInTransition : function() {
                    return this.inTransition;
                },

                getLastChangeTime : function() {
                    return this.lastChangeTime;
                },

                isState : function(stateName) {
                    return (stateName == this.state);
                }
            };
            return constructor;
        }();

}}

};

File: trunk/qp-java/qp-eclipse/src/salmir/util/StateMachine.java
package salmir.util;

import java.util.Set;
import java.util.HashSet;
import java.util.Map;
import java.util.HashMap;

//it is not thread-safe
//it is supposed all data passed to this object is valid
//It is the reason for all checks "assert" is used.

public class StateMachine {

  private Set<String> states = new HashSet<String>();
  //it is supposed that every transition only connects two states
  //sample: <start, {start, initialized, working}>
  private Map<String, String[]> transitions = new HashMap<String, String[]>();

  Object Owner;
  String initialState;
  String oldState;
  String state;

  boolean inTransition = false;
  String currentTransition = null;
  long lastChangeTime;


  public StateMachine(String[] stateNames, String initialStateName, String[][] transitionsData) {
      assert (stateNames != null);
      assert (initialStateName != null);
      assert (transitionsData != null);

      for (String state : stateNames) {
          states.add(state);
      }

      assert (states.contains(initialStateName));
      initialState = initialStateName;
                                              
      assert (transitionsData != null);

      for (int i = 0; i < transitionsData.length; i++) {
          //both states should be in the list of states
          assert (states.contains(transitionsData[i][1]));
          assert (states.contains(transitionsData[i][2]));

          transitions.put(transitionsData[i][0], transitionsData[i]);
      }

      this.oldState = null;
      state = initialState;
      lastChangeTime = System.currentTimeMillis();
  }

  public void startTransit(String transition) //throws Exception
  {
      assert (!inTransition): "The state machine is in transision: from '"
          					  + oldState + "' state by '" +currentTransition+"'";
      inTransition = true;
	  assert (transitions.containsKey(transition)): "The state machine does not contain this transition : " + transition;
      assert (state.equals(transitions.get(transition)[1])): "The from state of '" + transition + "' transition is not '" + state + "'";
      oldState = state;
      state = null;
      currentTransition = transition;
      lastChangeTime = System.currentTimeMillis();
  }

  public void endTransit() //throws Exception
  {
      assert (inTransition);
      state = transitions.get(currentTransition)[2];
      inTransition = false;
      currentTransition = null;
      lastChangeTime = System.currentTimeMillis();
  }

  public void doTransit(String transition)
  {
      startTransit(transition);
      endTransit();
  }

  public String getState() {
      return state;
  }

  public boolean isInTransition() {
      return inTransition;
  }

  public long getLastChangeTime() {
      return lastChangeTime;
  }

  public boolean isState(String stateName) {
      return stateName.equals(state); 
  }
}